src/Middleware/HttpLogging/src/ResponseBufferingStream.cs

```diff
@@ -111,49 +100,103 @@ private void CommonWrite(ReadOnlySpan<byte> span)
+            HttpLoggingMiddleware.LogResponseHeadersSync(_logContext, _options, _interceptors, _logger);
+            OnFirstWriteCore();
+        }
+    }
 
-            MediaTypeHelpers.TryGetEncodingForMediaType(_context.Response.ContentType, _encodings, out _encoding);
-            HeadersWritten = true;
+    private async ValueTask OnFirstWriteAsync()
+    {
+        if (!HeadersWritten)
+        {
+            // Log headers as first write occurs (headers locked now)
+            await HttpLoggingMiddleware.LogResponseHeadersAsync(_logContext, _options, _interceptors, _logger);
+            OnFirstWriteCore();
         }
     }
 
+    private void OnFirstWriteCore()
+    {
+        // The callback in LogResponseHeaders could disable body logging or adjust limits.
+        if (_logContext.LoggingFields.HasFlag(HttpLoggingFields.ResponseBody) && _logContext.ResponseBodyLogLimit > 0)
+        {
+            if (MediaTypeHelpers.TryGetEncodingForMediaType(_logContext.HttpContext.Response.ContentType,
+                _options.MediaTypeOptions.MediaTypeStates, out _encoding))
+            {
+                _logBody = true;
+                _limit = _logContext.ResponseBodyLogLimit;
+            }
+            else
+            {
+                _logger.UnrecognizedMediaType("response");
+            }
+        }
+
+        HeadersWritten = true;
+    }
+
     public void DisableBuffering()
     {
         _innerBodyFeature.DisableBuffering();
     }
 
-    public Task SendFileAsync(string path, long offset, long? count, CancellationToken cancellation)
+    public async Task SendFileAsync(string path, long offset, long? count, CancellationToken cancellation)
     {
-        OnFirstWrite();
-        return _innerBodyFeature.SendFileAsync(path, offset, count, cancellation);
+        await OnFirstWriteAsync();
+        await _innerBodyFeature.SendFileAsync(path, offset, count, cancellation);
     }
 
-    public Task StartAsync(CancellationToken token = default)
+    public async Task StartAsync(CancellationToken token = default)
     {
-        OnFirstWrite();
-        return _innerBodyFeature.StartAsync(token);
+        await OnFirstWriteAsync();
+        await _innerBodyFeature.StartAsync(token);
     }
 
     public async Task CompleteAsync()
     {
+        await OnFirstWriteAsync();
         await _innerBodyFeature.CompleteAsync();
     }
 
     public override void Flush()
     {
-        OnFirstWrite();
+        OnFirstWriteSync();
         base.Flush();
     }
 
-    public override Task FlushAsync(CancellationToken cancellationToken)
+    public override async Task FlushAsync(CancellationToken cancellationToken)
     {
-        OnFirstWrite();
-        return base.FlushAsync(cancellationToken);
+        await OnFirstWriteAsync();
+        await base.FlushAsync(cancellationToken);
+    }
+
+    public void LogResponseBody(double? duration)
+    {
+        if (_logBody)
```

Comment:
To be honest I haven't quite read the code for this class so I may be wrong, but by having this if statement does it mean that the `duration` get logged only if the response has a body?

Reply:
You're right. I just pushed a change to simplify how Duration is logged. E.g. it only happens once at the end of the response, and I don't combine it with the response body or headers.